Add landing page with table to Visualize app.

cjcenizal opened the pull request:
Addresses https://github.com/elastic/kibana/issues/9722

### Changes

- Refactor Visualize Wizard to use explicit controller names, instead of defining them dynamically.
- Update Visualize wizard UI.

### TODO

#### Now

- [x] Add icon-specific table column to UI Framework
- [x] Add kuiIcon variation for presenting a neutral icon
- [x] “Open” menu item should go back to home
- [x] Add column sorting
- [x] Fix visualization breadcrumbs to include Visualize app, and remove "Open" menu
- [x] Fix breadcrumbs to provide links back to each breadcrumb (and expect each breadcrumb link to be explicitly defined, maybe)
- [x] Adjust routes to support breadcrumbs representing progress through the Wizard
- [x] Add pagination to handle 1000s of visualizations

#### Later

- [ ] Onboarding content, when no visualizations exist
- [ ] Use lookup service to get icons, instead of relying on server
- [ ] Extract API service
- [ ] Use table and pagination directives and services

### Screenshots

![image](https://cloud.githubusercontent.com/assets/1238659/21412313/329bb7e4-c7a4-11e6-98f6-78b38b0e6524.png)

<img width="1423" alt="wizard_1" src="https://cloud.githubusercontent.com/assets/1238659/21410877/f3744e7e-c798-11e6-8ab2-3d5c9757f5c3.png">

<img width="1435" alt="wizard_2" src="https://cloud.githubusercontent.com/assets/1238659/21410878/f37580aa-c798-11e6-8015-ebf11bb7920b.png">

alt74 commented:
cool! @cjcenizal  instead of "+ new visualization" text in the button should we always use "Add New" instead? Use "Add New" for all tables where a user can create a new object (Graph, Timelion, etc).  @uboness  what do you think?

cjcenizal commented:
@alt74 That's a really interesting question. I think my preference as a user is to see the noun inside this button because it reinforces the domain of this view. Until we have a large "Visualizations" title, I feel like this view doesn't scream "Visualizations" at me, so anything we can do to reinforce the domain would be beneficial.

On the other hand, one advantage of a consistent label is that the buttons become more scalable -- we wouldn't need to worry about them growing too long and disrupting other controls inside the ToolBar. Can you think of others, @alt74 ?

Does the benefit of having the noun inside the button make sense? If so, then I propose we move forward by using the approach in this PR, but with a critical eye on each use case... if we discover a situation where the noun causes the button to disrupt other content in the ToolBar, then we can relabel them to have the more generic copy.

uboness commented:
I had the same feeling as @alt74... feels too repetitive indeed to have "Visualization" there. 

"+ Add New" also feels somewhat repetitive (is there an option to add something that is not new... not here at least"

Then we're left with "+ Add" do we really need both "+" and "Add" ?

I'd like to challenge having any text. "+" generally peaks for itself... just like a trash icon communicates "delete". I think that together with a tooltip, users will get used to this fairly quickly and you gain cleaner UI.

Also,

How many rows are we going to show in the table. Are we paginating at all here? We should. Now, if we don't have enough rows to require pagination and we hide the pagination buttons, should we just push the tool buttons ("+" in this case) all the way to the right? (right now it feels somewhat hanging there)

spalger commented on src/core_plugins/kibana/public/visualize/index.js:
```diff
@@ -24,7 +25,7 @@ uiRoutes
 })
 .when('/visualize', {
-  redirectTo: '/visualize/step/1'
+  redirectTo: '/visualize/landing'
```
Maybe we should pick a more descriptive route name. Maybe `/visualize/load` or `/visualize/index` or just `/visualize` since it's an index of the visualizations.

spalger replied:
Maybe we should pick a more descriptive route name. Maybe `/visualize/load`, `/visualize/index`, or just `/visualize` since it's an index of the visualizations.

spalger commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -6,24 +6,19 @@ import 'plugins/kibana/discover/saved_searches/saved_searches';
 /********/
 routes.when('/visualize/step/1', {
-  template: templateStep(1, require('plugins/kibana/visualize/wizard/step_1.html'))
+  template: require('plugins/kibana/visualize/wizard/step_1.html'),
```
Mind converting this to be an import statement?

spalger replied:
Mind converting this and the require below to be import statements?

spalger commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -6,24 +6,19 @@ import 'plugins/kibana/discover/saved_searches/saved_searches';
 routes.when('/visualize/step/1', {
-  template: templateStep(1, require('plugins/kibana/visualize/wizard/step_1.html'))
+  template: require('plugins/kibana/visualize/wizard/step_1.html'),
+  controller: 'VisualizeWizardStep1',
```
👍

spalger commented on src/ui_framework/components/button/_button.scss:
```diff
@@ -80,23 +80,23 @@
   }
 }
 
+/**
```
This should probably move up above line 70

spalger replied:
This should probably move up above line 70

cjcenizal replied:
I've found it's better to keep comments local to classes (even if they're duplicated), or else it can become really hard to keep track of them as the number of comments and classes grows.

spalger replied:
That's fine, but you moved this comment up from 106, so maybe it should be duplicated for line 76 and 106?

cjcenizal replied:
Good catch, but I think it already is duplicated for those lines.

ppisljar commented:
Regarding the button, i feel the noun inside the button is useful for the first couple of times you are using the application. After that its just taking extra space. I don't think we should have that button have the noun inside just for the sake of first time users, tooltip is fine enough for that.

I have a couple of other questions:
- list of visualizations as the first page .... i am not sure how kibana is used the most, but i am guessing you go to `Visualize` when you want to create a new chart. If that's the case then we just moved it one click further. On the other way this is more similar to how applications normally function (get list of items for some data type where you can delete, edit or add them)
- going for fixed-width center positioned design ... currently it looks super weird as we are switching from full width to fixed width design. If we plan to migrate whole kibana to fixed width then i guess it'll become better at some point. Note that this change doesn't just affect visualize, but same happens when you go to saved objects in management. Maybe its not that noticable on a small screen but on UHD you get this:

![vis_des_1](https://cloud.githubusercontent.com/assets/13629809/21452725/37ae9cb8-c909-11e6-9bf5-7681260e25c0.png)

and index selection screen is not the same width as the first two (but thats probably a bug or something went wrong with merge, as in CJs screenshots above it looks ok.

![vis_des_2](https://cloud.githubusercontent.com/assets/13629809/21452747/51d5a906-c909-11e6-9bb2-38f895f7d8bf.png)

ppisljar commented:
for a short time you see a narrower table displayed (with no data in it) ... should probably be hidden with ng-cloak, a loader or at least make the empty table the same width as the filled one to prevent annoying flicker


![ibana20161223121848](https://cloud.githubusercontent.com/assets/13629809/21452847/12cfb516-c90a-11e6-8f38-6ce0224bf11a.gif)

ppisljar commented:
few more ideas:
- i would go without type column
- i would put edit object icon on the right (something like in google inbox)
- delete/ create button should be right aligned ?
- hover on rows ?
- add additional buttons next to edit object (like share, duplicate, delete)

uboness commented:
> going for fixed-width center positioned design ... currently it looks super weird as we are switching from full width to fixed width design. If we plan to migrate whole kibana to fixed width then i guess it'll become better at some point. Note that this change doesn't just affect visualize, but same happens when you go to saved objects in management. Maybe its not that noticable on a small screen but on UHD you get this when you go to visualize (every other kibana page is full width):

agree... lets try 2 other options - a relative width (say 75% of the width centralized) and a full width

epixa commented:
I propose a third option: a max-width left justified.  No amount of wrangling is going to change the fact that so long as we have a sidebar in place, Kibana is left justified.  Centering content inside it just seems awkward.

cjcenizal commented:
## Comments from UI Review

### Wizard improvements

* The Wizard is tough to scan because it takes up so much space.
* Usually you know what you want your visualization to look like, but you may not know what it's called. It would help to be able to see a preview of what each type looks like.

### Table

* The column titles get a little lost.
* "Visualization" column could be renamed to just "Name".
* From using PagerDuty and GitHub, it seems like a green "Add" button would be more intuitive.
* The number of selected items gets lost on the bottom.

### Table actions

* Users may want a "Copy URL" link.
* We could explore other actions, such as Sharing and Reporting functionality from within the table.

cjcenizal commented:
## Changes

* I address @spalger's routing and `import` syntax suggestions.
* I fixed the changing-width bug that @ppisljar noted.
* I removed the button labels, added tooltips, and moved them to the right side per @uboness's suggestion.
* I renamed the "Visualization" column to "Name" per @kevinkluge's suggestion.

I think this is the limit to the changes I want to make in this PR. Let's tackle the question of table width and alignment in separate iteration, when we can take all table views into account.

cjcenizal commented:
jenkins, test this

Stacey-Gammon commented:
It looks awesome @cjcenizal!  Really glad about this, I think people will be really excited especially when we have one for dashboard.

A couple comments:
  - I can get into a situation where the timepicker shows in the upper right corner even on the landing page.
![timepickershowsup](https://cloud.githubusercontent.com/assets/16563603/21508858/17e83a14-cc52-11e6-9c7a-6f57b9259eab.gif)

  - I don't like the 'edit object' link. There are two different editing experiences - the raw editing of saved objects, that is more dangerous, and our "guided" ui experience, which is what the user gets from clicking on the link.  I don't think the distinction is clear that the < /> edit link brings you to saved object management but clicking the link brings you to the visualization (which you can also edit).  Do we really want saved object management to be so easily accessible?  Seems like a "power user" feature.
<img width="259" alt="screen shot 2016-12-27 at 4 30 48 pm" src="https://cloud.githubusercontent.com/assets/16563603/21508895/713dd100-cc52-11e6-88f1-5b430f72efc6.png">

- The column headers aren't sortable.  Since name used to be sortable, this seems like a regression. Type sorting was never possible before but would be cool if it was simple to add.

 - Minor thought - I think it'd be cool if the row you were hovering over was highlighted a slightly different color.  Take it or leave it. :)

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.html:
```diff
@@ -0,0 +1,156 @@
+
+      <tbody>
+        <tr
+          ng-repeat="item in listingController.items track by item.id | orderBy:'title'"
```
Is this supposed to cause it to be ordered by title because that doesn't seem to be happening?  Or does it default to order by title if there is no item.id?  I'm unfamiliar with this syntax.

cjcenizal replied:
Good catch, this actually needs to be:

```
ng-repeat="item in listingController.items | orderBy:'title' track by item.id"
```

Relevant links:

* https://code.angularjs.org/1.4.14/docs/api/ng/filter/orderBy
* http://stackoverflow.com/questions/16455832/orderby-not-working-as-expected-angularjs

I'll make a PR to fix this in Dashboard too.

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.html:
```diff
@@ -0,0 +1,156 @@
+      </div>
+    </div>
+  </div>
+</div>
```
The previous table component had max 20 items shown at one time.  Will this show all of them? Will the page be able to handle many visualizations?  It's not paged, right, they are just all listed?  I wonder how many visualizations some of our biggest users have, order of 1,000s maybe?

ppisljar replied:
there is an infinite_scroll under ui/directives

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,83 @@
+    const selectedIds = selectedItems.map(item => item.id);
+
+    visualizationService.delete(selectedIds)
+    .then(fetchObjects)
```
Should the .then's be indented by two because it's a continuation of the top line?

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,83 @@
+
+  this.edit = function edit(item) {
+    const params = {
+      // TODO: Get this value from somewhere instead of hardcodign it.
```
minor, but hardcodign = hardcoding

Stacey-Gammon commented on src/ui_framework/components/button/_button.scss:
```diff
@@ -73,7 +73,7 @@
   &:hover:enabled,
   &:active:enabled {
-    color: #FFFFFF;
+    color: #FFFFFF !important; /* 1 */
```
I see this `/* 1 */` used in various places, what's its meaning/purpose?

ppisljar replied:
there are comments at the top of the file marked with 1. and 2.

Stacey-Gammon replied:
ahhh, cool, thanks for the explanation!

Stacey-Gammon commented on test/support/page_objects/visualize_page.js:
```diff
@@ -372,14 +369,7 @@ export default class VisualizePage {
   // this is for starting on the
```
Is this comment relevant? Opening a saved visualization doesn't seem to have anything to do with the create new visualization page.

ppisljar commented:
I agree with everything Stacey said above, but besides that I have nothing more to add. This looks A W E S O M E. great job CJ!

tbragin commented:
@cjcenizal What do you think about discussion in here about re-naming, re-arranging, and removing descriptions? https://github.com/elastic/kibana/pull/9725#issuecomment-272034444

cjcenizal commented:
@tbragin I think that should be a separate discussion. I think they're related concerns, but they're also uncoupled. If we can keep them that way then we'll be able to keep the relevant code changes atomic and the respective discussions more focused.

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.html:
```diff
@@ -116,18 +116,6 @@
-
-              <button
-                class="kuiMicroButton kuiTableRowHoverReveal"
-                ng-click="listingController.edit(item)"
```
The edit function still exists in the controller, should probably remove that too

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.html:
```diff
@@ -0,0 +1,144 @@
+        <button
+          class="kuiButton kuiButton--danger"
+          confirm-click="listingController.deleteSelectedItems()"
+          confirmation="Are you sure you want to delete the selected visualizations? This action is irreversible!"
```
Looks like this uses the native confirm dialog, we should use the new one.

Stacey-Gammon commented:
When there are no visualizations/dashboards, do you think we should we call out more the [+] button?  Kind of like do when there are no panels in a dashboard.

<img width="1229" alt="screen shot 2017-01-20 at 4 34 33 pm" src="https://cloud.githubusercontent.com/assets/16563603/22166305/66dcae40-df2e-11e6-97d9-afd602c9d0a6.png">

cjcenizal commented:
@Stacey-Gammon Great idea! I just created a PR for a prompt when there are no items in a table or list. Can you take a quick look and let me know what you think? https://github.com/elastic/kibana/pull/10016

tbragin commented:
@cjcenizal is there a separate ticket / PR you're tracking to improve the "select visualization type" page? I'd like to add to it things we discussed here, such as removing descriptions, improving icons, renaming, and re-ordering https://github.com/elastic/kibana/pull/9725#issuecomment-272034444

tbragin commented:
@cjcenizal I checked out the PR and a couple of questions -- you have column sorting listed for "later"... is that still in the context of this PR? Since we made the dashboard columns sortable in https://github.com/elastic/kibana/pull/10003 i think it should be. 

I think that adding created/modified dates to these tables as well as tables in Management (which appear to be part of this PR as well?) should be a priority. Perhaps a separate effort, but i'd be curious to see how it affects spacing/layout within these tables. 

cc: @epixa

cjcenizal commented:
@tbragin No issue exists yet for tracking the "select visualization type" changes you suggested. Could you please create one and tag it as "blocked" by this PR?

Also, I updated the todo checklist in the description per your comments. Thanks!

tbragin commented:
@cjcenizal finally created this related issue, per your request https://github.com/elastic/kibana/issues/10145

cjcenizal commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -29,9 +27,15 @@ module.controller('VisualizeWizardStep1', function ($scope, $route, kbnUrl, time
+
   $scope.visTypeUrl = function (visType) {
-    const baseUrl = visType.requiresSearch ? '#/visualize/step/2?' : '#/visualize/create?';
+    const baseUrl =
```
Does anyone think we need to add redirects from the old routes to the new routes? Could we be breaking any functionality, e..g bookmarks or third-party plugins, by changing these routes?

spalger replied:
We absolutely should if we can. What routes are changing?

cjcenizal replied:
* `#/visualize/step/1` => `/visualize/new`
* `#/visualize/step/2` => `/visualize/new/configure`

I'll add redirects here. Any suggestions on the best way to do that?

cjcenizal replied:
Done, using `redirectTo`.

cjcenizal commented:
@spalger @ppisljar @Stacey-Gammon Ready for another review.

* Adding pagination to both Visualize and Dashboard listing tables. Stacey, can you make sure Dashboard still works OK?
* I changed the routes for the Visualize wizard. Do these still make sense to everyone? Suggestions welcome for better route names, which are displayed in the breadcrumbs.
* Added sorting on table.
* Removed "Open" menu button from the Visualization view.

cjcenizal commented:
jenkins, test this

Stacey-Gammon commented on src/core_plugins/kibana/public/dashboard/listing/dashboard_listing.html:
```diff
@@ -53,12 +49,27 @@
         </a>
       </div>
+
+      <div class="kuiToolBarSection">
```
Awesome!  Only one issue is that if you select all, looks like it selects everything, not just the first page, so if you do select all and delete, it'll delete everything.  I don't think this is what we want.

spalger replied:
I agree with @Stacey-Gammon. The way I see it, you shouldn't be able to select items that are not visible. This would also mean that changing the page would clear previously selected items.

cjcenizal replied:
Great catch! Thanks, will update.

Stacey-Gammon commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.html:
```diff
@@ -0,0 +1,172 @@
+    </div>
+
+    <!-- NoResults -->
+    <div class="kuiPanel kuiPanel--centered" ng-if="!listingController.items.length">
```
Use the new `Prompt For Items` section?  Also for some reason this isn't coming out as styled like the dashboard one, I think kuiNoResults should be kuiNoItems.

<img width="1126" alt="screen shot 2017-02-06 at 10 00 38 am" src="https://cloud.githubusercontent.com/assets/16563603/22652108/21a759a6-ec53-11e6-84c0-391a8d0943a1.png">

Stacey-Gammon commented:
Love the new routes and the new look/feel of creating a visualization.  So much cleaner!

spalger commented on src/core_plugins/kibana/public/visualize/editor/editor.html:
```diff
@@ -3,16 +3,14 @@
-          ng-bind="savedVis.lastSavedTitle"
-        ></span>
-      </div>
+        title="getVisualizationTitle()"
```
this should probably be `page-title`

spalger replied:
`title` isn't used by the breadcrumbs so the navbar no longer has the item title in it.

cjcenizal replied:
Great catch, thanks!

spalger commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,167 @@
+import _ from 'lodash';
+
+export function VisualizeListingController(
+  $filter,
```
Would you mind trying `const $filter = $injector.get('$filter')` here and below? It's a part of https://github.com/elastic/kibana/pull/9495 that, among other things, prevents argument lists like this

cjcenizal replied:
Sure.

spalger commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,167 @@
+    return `#/visualize/edit/${item.id}`;
+  };
+
+  $scope.$watch(() => this.filter, () => {
```
I think this belongs with the rest of the initialization logic up top (or even better in a constructor)

spalger commented on src/core_plugins/kibana/public/visualize/visualize_constants.js:
```diff
@@ -0,0 +1,7 @@
+export const VisualizeConstants = {
+  LANDING_PAGE_URL: '/visualize',
```
The way these `*_URL` variables are used implies that `*_PATHNAME` would be a more appropriate suffix (The code assumes they will not have query strings, hosts, ports, etc.).

spalger commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -77,8 +89,11 @@ module.controller('VisualizeWizardStep2', function ($route, $scope, timefilter,
 
     if (addToDashMode) {
-      return `#/visualize/create?${DashboardConstants.ADD_VISUALIZATION_TO_DASHBOARD_MODE_PARAM}&type=${type}&indexPattern=${pattern}`;
+      return `#${VisualizeConstants.CREATE_URL}` +
```
There is a lot of this going on, I don't want to comment on it everywhere, but I'd like to reiterate my opinion that dynamically creating urls with string concatenation is a recipe for hard to trace bugs. Ideally we would always use a bonafide formatter like `url.format()`

cjcenizal replied:
Recorded in https://github.com/elastic/kibana/issues/10199

spalger commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,167 @@
+  };
+
+  this.isItemChecked = function isItemChecked(item) {
+    return selectedItems.indexOf(item) !== -1;
```
`Array#includes()`?

cjcenizal replied:
IE doesn't support this yet. Does it get shimmed?

spalger replied:
Yep, we use the babel-polyfill so we get all of the new Array and String methods, along with Maps and Sets and such

ppisljar replied:
i hope it does as we do use it in many other locations.

spalger commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -0,0 +1,167 @@
+  }];
+
+  this.getSortProperty = function getSortProperty() {
+    return _.find(this.sortProperties, property => property.isSelected);
```
No need for `_.find()`, `Array#find()` should do the trick

spalger commented:
Most of my comments are style/optional/discretionary items, but the title missing from the breadcrumbs and potentially the ability to select non-visible items are things I think we need to fix before merging.

cjcenizal commented on src/ui/public/pager/pager_service.factory.js:
```diff
@@ -0,0 +1,12 @@
+
+const app = uiModules.get('kibana');
+
+app.factory('pagerService', () => {
```
I think we should name this `pagerFactory` instead, which is a much clearer indication IMO that this object follows the factory pattern (i.e. it exposes `create` methods).

cjcenizal commented:
@Stacey-Gammon @spalger I've addressed your feedback. Would you mind taking another look? Please see the commit log for detailed notes on the changes.

A couple of the more visual changes:

## Use NoItems and PromptForItems components

![image](https://cloud.githubusercontent.com/assets/1238659/22663662/4b27313e-ec62-11e6-891d-cfedea224b86.png)
